File: app/src/main/java/demos/util/SimpleAnimator.java
package demos.util;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.Interpolator;

/**
 * Created by victor on 16-7-7
 */
public final class SimpleAnimator {
    private View targetView;
    private long duration = 300L;
    private String property = "x";
    private Number[] values;
    private Interpolator interpolator = new AccelerateDecelerateInterpolator();
    private AnimatorListener listener;
    private ValueAnimator.AnimatorUpdateListener updateListener;

    private SimpleAnimator() {
    }

    private SimpleAnimator(View targetView) {
        this.targetView = targetView;
    }

    public static SimpleAnimator create(View targetView) {
        return new SimpleAnimator(targetView);
    }

    public SimpleAnimator values(Number... values) {
        this.values = values;
        return this;
    }

    public SimpleAnimator duration(long duration) {
        this.duration = duration;
        return this;
    }

    public SimpleAnimator property(String property) {
        this.property = property;
        return this;
    }

    public SimpleAnimator interpolator(Interpolator interpolator) {
        this.interpolator = interpolator;
        return this;
    }

    public SimpleAnimator listener(AnimatorListener listener) {
        this.listener = listener;
        return this;
    }

    public SimpleAnimator updateListener(ValueAnimator.AnimatorUpdateListener updateListener) {
        this.updateListener = updateListener;
        return this;
    }

    public void go() {
        if (targetView == null)
            return;
        if (values.length < 2) {
            throw new IllegalArgumentException("values起码得两个值啊！");
        }
        ValueAnimator animator;
        if (property.contentEquals("height") || property.contentEquals("width")) {
            final int[] intValues = new int[this.values.length];
            for (int i = 0; i < this.values.length; i++) {
                intValues[i] = (int) this.values[i];
            }
            final ViewGroup.LayoutParams lp = targetView.getLayoutParams();
            animator = new ValueAnimator();
            animator.setIntValues(intValues);
            animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
                @Override
                public void onAnimationUpdate(ValueAnimator animation) {
                    int value = (int) animation.getAnimatedValue();
                    if (property.contentEquals("width")) {
                        lp.width = value;
                    }
                    if (property.contentEquals("height")) {
                        lp.height = value;
                    }
                    targetView.setLayoutParams(lp);
                }
            });
            animator.addListener(new AnimatorListener() {
                @Override
                public void onAnimationStart(Animator animation) {
                    targetView.setVisibility(View.VISIBLE);
                }

                @Override
                public void onAnimationEnd(Animator animation) {
                    if (intValues[intValues.length - 1] == 0) {
                        targetView.setVisibility(View.INVISIBLE);
                    }
                }
            });
        } else {
            final float[] floatValues = new float[this.values.length];
            for (int i = 0; i < this.values.length; i++) {
                floatValues[i] = (float) this.values[i];
            }
            animator = ObjectAnimator.ofFloat(this.targetView, property, floatValues);

            animator.addListener(new AnimatorListener() {
                @Override
                public void onAnimationStart(Animator animation) {
                    targetView.setVisibility(View.VISIBLE);
                }

                @Override
                public void onAnimationEnd(Animator animation) {
                    if (property.contentEquals("alpha") && floatValues[floatValues.length - 1] == 0f) {
                        targetView.setVisibility(View.INVISIBLE);
                    }
                }
            });
        }
        animator.setDuration(duration);
        animator.setInterpolator(interpolator);
        if (listener != null) {
            animator.addListener(listener);
        }
        if (updateListener != null) {
            animator.addUpdateListener(updateListener);
        }
        animator.start();
    }

    public static abstract class AnimatorListener implements Animator.AnimatorListener {

        @Override
        public void onAnimationStart(Animator animation) {

        }

        @Override
        public void onAnimationEnd(Animator animation) {

        }

        @Override
        public void onAnimationCancel(Animator animation) {

        }

        @Override
        public void onAnimationRepeat(Animator animation) {

        }
    }
}

File: app/src/main/kotlin/demos/util/SimpleAnimator.kt
package demos.util

import android.animation.Animator
import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.view.View
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.Interpolator

/**
 * Created by victor on 16-7-7.
 */
class SimpleAnimator(private val targetView: View?) {
    private var duration = 300L
    private var property = ""
    private var values: Array<out Number> = arrayOf()
    private var interpolator: Interpolator = AccelerateDecelerateInterpolator()
    private var listener: AnimatorListener? = null
    private var updateListener: ValueAnimator.AnimatorUpdateListener? = null

    fun values(vararg values: Number): SimpleAnimator {
        this.values = values
        return this
    }

    fun duration(duration: Long): SimpleAnimator {
        this.duration = duration
        return this
    }

    fun property(property: String): SimpleAnimator {
        this.property = property
        return this
    }

    fun interpolator(interpolator: Interpolator): SimpleAnimator {
        this.interpolator = interpolator
        return this
    }

    fun listener(listener: AnimatorListener): SimpleAnimator {
        this.listener = listener
        return this
    }

    fun updateListener(updateListener: ValueAnimator.AnimatorUpdateListener): SimpleAnimator {
        this.updateListener = updateListener
        return this
    }

    fun go() {
        if (targetView == null)
            return
        if (values.size < 2) {
            throw IllegalArgumentException("values起码得两个值啊！")
        }
        val animator: ValueAnimator
        if (property.contentEquals("height") || property.contentEquals("width")) {
            val intValues = arrayListOf<Int>()
            this.values.forEach { intValues.add(it.toInt()) }
            val lp = targetView.layoutParams
            animator = ValueAnimator()
            animator.setIntValues(*(intValues.toIntArray()))
            animator.addUpdateListener {
                when (property) {
                    "width" -> lp.width = it.animatedValue as Int
                    "height" -> lp.height = it.animatedValue as Int
                }
                targetView.layoutParams = lp
            }
            animator.addListener(object : AnimatorListener() {
                override fun onAnimationEnd(animation: Animator?) {
                    if (intValues[intValues.lastIndex] == 0) {
                        targetView.visibility = View.INVISIBLE
                    }
                }

                override fun onAnimationStart(animation: Animator?) {
                    targetView.visibility = View.VISIBLE
                }
            })
        } else {
            val floatValues = arrayListOf<Float>()
            this.values.forEach { floatValues.add(it.toFloat()) }
            animator = ObjectAnimator.ofFloat(this.targetView, property, *(floatValues.toFloatArray()))
            animator.addListener(object : AnimatorListener() {
                override fun onAnimationEnd(animation: Animator?) {
                    if (property.contentEquals("alpha") && floatValues[floatValues.lastIndex] == 0f) {
                        targetView.visibility = View.INVISIBLE
                    }
                }

                override fun onAnimationStart(animation: Animator?) {
                    targetView.visibility = View.VISIBLE
                }
            })
        }
        animator.duration = duration
        animator.interpolator = interpolator
        listener?.let { animator.addListener(listener) }
        updateListener?.let { animator.addUpdateListener(updateListener) }
        animator.start()
    }

    open class AnimatorListener : Animator.AnimatorListener {
        override fun onAnimationRepeat(animation: Animator?) {
        }

        override fun onAnimationEnd(animation: Animator?) {
        }

        override fun onAnimationCancel(animation: Animator?) {
        }

        override fun onAnimationStart(animation: Animator?) {
        }
    }
}